scripts/linters/custom_eslint_checks/rules/comment-checker.js

```diff
@@ -0,0 +1,134 @@
+
+      // We allow comments to not have a terminating punctuation if any of the
+      // below phrases appears at the beginning of the comment.
+      // Example: // eslint - disable max-len
+      // This comment will be excluded from this check.
+      const allowedStartPhrases = [
+        '@ts-expect-error', '@ts-ignore', '--params', 'eslint-disable',
+        'eslint-enable'];
+
+      // We allow comments to not have a terminating punctuation if any of the
+      // below phrases appears in the last word of a comment.
+      // Example: // Ref: https://some.link.com
+      // This comment will be excluded from this check.
+      const allowedEndPhrases = ['http://', 'https://'];
+
+      // Check if any of the allowed starting phrase is present in
+      // comment and exclude that line from check.
+      var allowedStartPhrasePresent = false;
+      allowedStartPhrases.forEach((phrase) => {
+        if (comment.value.split(' ')[1].startsWith(phrase)) {
+          allowedStartPhrasePresent = true;
+        }
+      });
+
+      if (allowedStartPhrasePresent) {
+        return true;
+      }
+
+      // Check if any of the allowed ending phrase is present
+      // in comment and exclude that line from check. Used 'includes'
+      // instead of 'startsWith' because we have some comments
+      // with urls inside the quotes.
+      // Example: 'https://oppia.org'
+      var allowedEndPhrasePresent = false;
+      allowedEndPhrases.forEach((phrase) => {
+        var splittedComment = comment.value.split(' ');
+        if (splittedComment[splittedComment.length - 1].includes(phrase)) {
+          allowedEndPhrasePresent = true;
+        }
+      });
+
+      if (allowedEndPhrasePresent) {
+        return true;
+      }
+
+      // Check that the comments end with the proper punctuation.
+      if (comment.type === 'Block') {
+        var lastCharIsInvalid = !(allowedTerminatingPunctuations.includes(
+          comment.value[comment.value.length - 3]));
+      } else {
+        var lastCharIsInvalid = !(allowedTerminatingPunctuations.includes(
+          comment.value[comment.value.length - 1]));
+      }
+      if (lastCharIsInvalid) {
+        context.report({
+          comment,
+          loc: comment.loc,
+          messageId: 'invalidPunctuation'
+        });
+      }
+    };
+
+    return {
+      Program() {
+        var comments = [];
+        const commentTokens = sourceCode.getAllComments();
+        commentTokens.forEach((token) => {
+          if (token.type !== 'Shebang') {
+            comments.push(token);
+          }
+        });
+
+        for (var i = 1; i < comments.length - 1; i++) {
+          var prevComment = comments[i - 1];
+          var comment = comments[i];
+          var nextComment = comments[i + 1];
+
+          if (comment.value.endsWith('-') &&
+            !(nextComment.loc.start.line - 1 === comment.loc.start.line &&
+              prevComment.loc.start.line + 1 === comment.loc.start.line)) {
```

Comment:
Can you please explain this check? [ALso add a code comment as it looks complicated.]